Py_UNICODE_ISALNUM and similar should take Py_UCS4 but take Py_UNICODE instead

See e.g. https://github.com/pypy/pypy/blob/a964c5a891c8168fc3e3128bad6257546d362455/pypy/module/cpyext/unicodeobject.py#L131-L132

But in the CPython header:
https://github.com/python/cpython/blob/590319072773bd6cdcca655c420d3adb84838e96/Include/cpython/unicodeobject.h#L691

See also https://github.com/ngoldbaum/pypy-bug which contains a runnable example based on a cython script to test for the bug. When I build it on 64 bit windows pypy3.10, I see in the compiler output:

```
test_cython.c(2362): warning C4304: 'function': truncation from 'Py_UCS4' to 'wchar_t'
```

Discovered in a new test added to NumPy: https://github.com/numpy/numpy/pull/25791

--- a/pypy/module/cpyext/test/test_typeobject.py
+++ b/pypy/module/cpyext/test/test_typeobject.py
@@ -2064,6 +2064,43 @@ class AppTestSlots(AppTestCpythonExtensionBase):
         res = module.test_fastcalldict(pyfunc, (1, ), {"arg2": 2})
         assert res == [1, 2]
 
+    def test_getset_doc(self):
+        module = self.import_extension("foo", [
+            ("getset_type", "METH_NOARGS",
+            """
+            PyObject *ret = PyType_FromSpec(&test_spec);
+            return ret;
+            """)], prologue="""
+            typedef struct {
+                PyObject_HEAD
+            } Test;
+
+            static int test_init(PyObject *self, PyObject *args, PyObject *kwd) { return 0; }
+            static PyObject *test_getter(PyObject *self, void* payload) { return PyLong_FromLong(123); }
+
+            static PyGetSetDef test_getset[] = {
+                { "prop", test_getter, NULL, "A docstring", NULL },
+                {NULL, NULL, NULL, NULL, NULL }
+            };
+
+            static PyType_Slot test_slots[] = {
+                { Py_tp_init, test_init },
+                { Py_tp_getset, test_getset },
+                { 0, NULL }
+            };
+
+            static PyType_Spec test_spec = {
+                .name = "Test",
+                .flags = Py_TPFLAGS_DEFAULT,
+                .slots = test_slots,
+                .basicsize = (int) sizeof(Test),
+                .itemsize = 0
+            };
+            """)
+
+        test = module.getset_type()
+        assert test.prop.__doc__ == "A docstring"
+        
 
 class AppTestHashable(AppTestCpythonExtensionBase):
     def test_unhashable(self):

--- a/pypy/module/cpyext/typeobject.py
+++ b/pypy/module/cpyext/typeobject.py
@@ -58,8 +58,7 @@ class W_GetSetPropertyEx(GetSetProperty):
         self.getset = getset
         self.w_type = w_type
         doc = fset = fget = fdel = None
-        if doc:
-            # XXX dead code?
+        if getset.c_doc:
             doc = rffi.constcharp2str(getset.c_doc)
         if getset.c_get:
             fget = GettersAndSetters.getter.im_func

--- a/pypy/module/cpyext/unicodeobject.py
+++ b/pypy/module/cpyext/unicodeobject.py
@@ -1,6 +1,6 @@
 from rpython.rtyper.lltypesystem import rffi, lltype
 from rpython.rlib import rstring
-from rpython.rlib.rarithmetic import widen, r_uint
+from rpython.rlib.rarithmetic import widen, r_uint, r_uint32, intmask
 from rpython.rlib import rstring, rutf8
 from rpython.tool.sourcetools import func_renamer
 
@@ -287,127 +287,127 @@ def set_compact(py_obj, value):
     get_state(py_obj).c_compact = cts.cast('unsigned char', value)
 
 
-@cpython_api([Py_UNICODE], rffi.INT_real, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], rffi.INT_real, error=CANNOT_FAIL)
 def Py_UNICODE_ISSPACE(space, ch):
     """Return 1 or 0 depending on whether ch is a whitespace character."""
-    ch = ord(ch)
+    ch = widen(ch)
     if ch >= rutf8.MAXUNICODE:
         return 0
     return unicodedb.isspace(ch)
 
-@cpython_api([Py_UNICODE], rffi.INT_real, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], rffi.INT_real, error=CANNOT_FAIL)
 def Py_UNICODE_ISALPHA(space, ch):
     """Return 1 or 0 depending on whether ch is an alphabetic character."""
-    ch = ord(ch)
+    ch = widen(ch)
     if ch >= rutf8.MAXUNICODE:
         return 0
     return unicodedb.isalpha(ch)
 
-@cpython_api([Py_UNICODE], rffi.INT_real, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], rffi.INT_real, error=CANNOT_FAIL)
 def Py_UNICODE_ISALNUM(space, ch):
     """Return 1 or 0 depending on whether ch is an alphanumeric character."""
-    ch = ord(ch)
+    ch = widen(ch)
     if ch >= rutf8.MAXUNICODE:
         return 0
     return unicodedb.isalnum(ch)
 
-@cpython_api([Py_UNICODE], rffi.INT_real, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], rffi.INT_real, error=CANNOT_FAIL)
 def Py_UNICODE_ISLINEBREAK(space, ch):
     """Return 1 or 0 depending on whether ch is a linebreak character."""
-    ch = ord(ch)
+    ch = widen(ch)
     if ch >= rutf8.MAXUNICODE:
         return 0
     return unicodedb.islinebreak(ch)
 
-@cpython_api([Py_UNICODE], rffi.INT_real, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], rffi.INT_real, error=CANNOT_FAIL)
 def Py_UNICODE_ISDECIMAL(space, ch):
     """Return 1 or 0 depending on whether ch is a decimal character."""
-    ch = ord(ch)
+    ch = widen(ch)
     if ch >= rutf8.MAXUNICODE:
         return 0
     return unicodedb.isdecimal(ch)
 
-@cpython_api([Py_UNICODE], rffi.INT_real, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], rffi.INT_real, error=CANNOT_FAIL)
 def Py_UNICODE_ISDIGIT(space, ch):
     """Return 1 or 0 depending on whether ch is a digit character."""
-    ch = ord(ch)
+    ch = widen(ch)
     if ch >= rutf8.MAXUNICODE:
         return 0
     return unicodedb.isdigit(ch)
 
-@cpython_api([Py_UNICODE], rffi.INT_real, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], rffi.INT_real, error=CANNOT_FAIL)
 def Py_UNICODE_ISNUMERIC(space, ch):
     """Return 1 or 0 depending on whether ch is a numeric character."""
-    ch = ord(ch)
+    ch = widen(ch)
     if ch >= rutf8.MAXUNICODE:
         return 0
     return unicodedb.isnumeric(ch)
 
-@cpython_api([Py_UNICODE], rffi.INT_real, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], rffi.INT_real, error=CANNOT_FAIL)
 def Py_UNICODE_ISLOWER(space, ch):
     """Return 1 or 0 depending on whether ch is a lowercase character."""
-    ch = ord(ch)
+    ch = widen(ch)
     if ch >= rutf8.MAXUNICODE:
         return 0
     return unicodedb.islower(ch)
 
-@cpython_api([Py_UNICODE], rffi.INT_real, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], rffi.INT_real, error=CANNOT_FAIL)
 def Py_UNICODE_ISUPPER(space, ch):
     """Return 1 or 0 depending on whether ch is an uppercase character."""
-    ch = ord(ch)
+    ch = widen(ch)
     if ch >= rutf8.MAXUNICODE:
         return 0
     return unicodedb.isupper(ch)
 
-@cpython_api([Py_UNICODE], rffi.INT_real, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], rffi.INT_real, error=CANNOT_FAIL)
 def Py_UNICODE_ISTITLE(space, ch):
     """Return 1 or 0 depending on whether ch is a titlecase character."""
-    ch = ord(ch)
+    ch = intmask(ch)
     if ch >= rutf8.MAXUNICODE:
         return 0
     return unicodedb.istitle(ch)
 
-@cpython_api([Py_UNICODE], Py_UNICODE, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], Py_UCS4, error=CANNOT_FAIL)
 def Py_UNICODE_TOLOWER(space, ch):
     """Return the character ch converted to lower case."""
-    val = ord(ch)
+    val = intmask(ch)
     if val >= rutf8.MAXUNICODE:
         return ch
-    return unichr(unicodedb.tolower(val))
+    return r_uint32(unicodedb.tolower(val))
 
-@cpython_api([Py_UNICODE], Py_UNICODE, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], Py_UCS4, error=CANNOT_FAIL)
 def Py_UNICODE_TOUPPER(space, ch):
     """Return the character ch converted to upper case."""
-    val = ord(ch)
+    val = intmask(ch)
     if val >= rutf8.MAXUNICODE:
         return ch
-    return unichr(unicodedb.toupper(val))
+    return r_uint32(unicodedb.toupper(val))
 
-@cpython_api([Py_UNICODE], Py_UNICODE, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], Py_UCS4, error=CANNOT_FAIL)
 def Py_UNICODE_TOTITLE(space, ch):
     """Return the character ch converted to title case."""
-    val = ord(ch)
+    val = intmask(ch)
     if val >= rutf8.MAXUNICODE:
         return ch
-    return unichr(unicodedb.totitle(val))
+    return r_uint32(unicodedb.totitle(val))
 
-@cpython_api([Py_UNICODE], rffi.INT_real, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], rffi.INT_real, error=CANNOT_FAIL)
 def Py_UNICODE_TODECIMAL(space, ch):
     """Return the character ch converted to a decimal positive integer.  Return
     -1 if this is not possible.  This macro does not raise exceptions."""
-    ch = ord(ch)
-    if ch >= rutf8.MAXUNICODE:
+    val = intmask(ch)
+    if val >= rutf8.MAXUNICODE:
         return -1
     try:
-        return unicodedb.decimal(ch)
+        return unicodedb.decimal(val)
     except KeyError:
         return -1
 
-@cpython_api([Py_UNICODE], rffi.INT_real, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], rffi.INT_real, error=CANNOT_FAIL)
 def Py_UNICODE_TODIGIT(space, ch):
     """Return the character ch converted to a single digit integer. Return -1 if
     this is not possible.  This macro does not raise exceptions."""
-    ch = ord(ch)
+    ch = widen(ch)
     if ch >= rutf8.MAXUNICODE:
         return -1
     try:
@@ -415,11 +415,11 @@ def Py_UNICODE_TODIGIT(space, ch):
     except KeyError:
         return -1
 
-@cpython_api([Py_UNICODE], rffi.DOUBLE, error=CANNOT_FAIL)
+@cpython_api([Py_UCS4], rffi.DOUBLE, error=CANNOT_FAIL)
 def Py_UNICODE_TONUMERIC(space, ch):
     """Return the character ch converted to a double. Return -1.0 if this is not
     possible.  This macro does not raise exceptions."""
-    ch = ord(ch)
+    ch = widen(ch)
     if ch >= rutf8.MAXUNICODE:
         return -1.0
     try:
@@ -1228,7 +1228,7 @@ def PyUnicode_TransformDecimalToASCII(space, s, size):
         ch = s[i]
         ordch = ord(ch)
         if ordch > 127:
-            decimal = Py_UNICODE_TODECIMAL(space, ch)
+            decimal = Py_UNICODE_TODECIMAL(space, ordch)
             decimal = rffi.cast(lltype.Signed, decimal)
             if decimal >= 0:
                 ordch = ord('0') + decimal

